src/workflow_tasks.rs

```diff
@@ -535,10 +580,15 @@ impl WorkflowTaskManager {
                 }
             }
 
-            // Evict run id if cache is full. Non-sticky will always evict.
-            let maybe_evicted = self.cache_manager.lock().insert(run_id);
-            if let Some(evicted_run_id) = maybe_evicted {
-                self.evict_run(&evicted_run_id);
+            if !just_evicted {
+                // Evict run id if cache is full. Non-sticky will always evict.
+                let maybe_evicted = self.cache_manager.lock().insert(run_id);
+                if let Some(evicted_run_id) = maybe_evicted {
+                    // Don't try to re-evict something we just evicted
+                    if evicted_run_id != run_id || !just_evicted {
+                        self.request_eviction(&evicted_run_id);
+                    }
+                }
```

Comment:
This should probably be refactored into a method `self.touch_and_maintain_cache(run_id)` or something like that.